### PR TITLE
[huggingface_pytorch, huggingface_tensorflow] Add support for speech & vision

### PR DESCRIPTION
**Pull request by philschmid**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [x] Revision A: `sagemaker_remote_tests = "standard"`
- [x] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [x] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [x] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [x] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [x] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [x] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [x] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by saimidu** [comment]
@ydaiming Do you think it would be okay to upgrade the numpy version used on the PyTorch 1.10.2 Hugging Face dockerfile to numpy 1.22.3 (which is currently the latest 1.22 version) in order to fix the failing test? Would upgrading numpy break PyTorch functionality on this DLC image?

File: test/sagemaker_tests/huggingface/inference/integration/sagemaker/test_speech_model.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    except Exception as e:
+        dump_logs_from_cloudwatch(e, region)
+        raise
+
+
+@pytest.mark.model("wav2vec2")
+@pytest.mark.processor("gpu")
+@pytest.mark.gpu_test
+def test_speech_model_gpu(sagemaker_session, framework_version, ecr_image, instance_type, region):
+    instance_type = instance_type or "ml.p2.xlarge"
```
Comment (saimidu) [inline test/sagemaker_tests/huggingface/inference/integration/sagemaker/test_speech_model.py:42]: Please change the "default" instance type in this statement to `ml.p3.2xlarge` instead of `ml.p2.xlarge`, as p2 instance types are not compatible with certain CUDA and GPU-driver versions.

File: test/sagemaker_tests/huggingface/inference/integration/sagemaker/test_vision_model.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    except Exception as e:
+        dump_logs_from_cloudwatch(e, region)
+        raise
+
+
+@pytest.mark.model("wav2vec2")
+@pytest.mark.processor("gpu")
+@pytest.mark.gpu_test
+def test_vision_model_gpu(sagemaker_session, framework_version, ecr_image, instance_type, region):
+    instance_type = instance_type or "ml.p2.xlarge"
```
Comment (saimidu) [inline test/sagemaker_tests/huggingface/inference/integration/sagemaker/test_vision_model.py:42]: Please change the "default" instance type in this statement to `ml.p3.2xlarge` instead of `ml.p2.xlarge`, as p2 instance types are not compatible with certain CUDA and GPU-driver versions.

File: huggingface/tensorflow/inference/docker/2.6/py3/cu112/Dockerfile.gpu [hunk 3]
```diff
@@ -42,6 +42,8 @@ RUN apt-get update \
     unzip \
     git \
     libnuma1 \
+    libsndfile1-dev \
+    ffmpeg \
```
Comment (saimidu) [inline huggingface/tensorflow/inference/docker/2.6/py3/cu112/Dockerfile.gpu:46]: These seem to be installed only on the inference dockerfiles, and not the training dockerfiles. Is this intentional?

We should have some kind of validation checks to ensure that hugging face DLCs for TF>=2.6 and PT>=1.10 have these libs installed, as well as tests to validate that the audio and vision transformers packages are installed.

Example: https://github.com/aws/deep-learning-containers/blob/master/test/dlc_tests/sanity/test_utility_installation.py

Comment (philschmid) [reply huggingface/tensorflow/inference/docker/2.6/py3/cu112/Dockerfile.gpu:46]: Yes they are used in the [transformers.pipeline](https://github.com/huggingface/transformers/blob/b320d87eceb369ea22d5cd73866499851cb2cca3/src/transformers/pipelines/audio_utils.py#L8).

Comment (saimidu) [reply huggingface/tensorflow/inference/docker/2.6/py3/cu112/Dockerfile.gpu:46]: Thanks! If they are necessary to pass the newly added audio and vision tests, then please ignore this comment.

**Comment by ydaiming** [comment]
I think it's okay to upgrade the `numpy` version.

PyTorch doesn't have a hard dependency on `numpy` version.

**Comment by arjkesh** [comment]
@philschmid - numpy version needs to be bumped even higher to remove CVEs in inference image. Is it possible to use 1.22.3?

**Comment by Qingzi-Lan** [comment]
`pip install numpy>=1.22.3` seems not working. I double checked it locally. however, `pip install numpy==1.22.3` works
I'd like to reconsider to upgrade numpy from `1.21.5` to `1.22.3`. I see dependency like this:

```
numpy ->
numba==0.55.1  # numpy<1.22,>=1.18
    librosa==0.9.1  # numba>=0.45.1
      sagemaker-huggingface-inference-toolkit==2.0.0  # librosa
    resampy==0.2.2  # numba>=0.32
      librosa==0.9.1  # resampy>=0.2.2
        sagemaker-huggingface-inference-toolkit==2.0.0  # librosa
```

then `numpy 1.21.5` is the highest version we can have in this case. I can add CVE to allowlist since we can't do more on it.

**Comment by Qingzi-Lan** [comment]
@philschmid the better practice is to update `sagemaker-huggingface-inference-toolkit` I think. then we can eliminated the CVE then

**Comment by philschmid** [comment]
> @philschmid the better practice is to update `sagemaker-huggingface-inference-toolkit` dependency I think. then we can eliminated the CVE

you mean pin `numpy` to `1.21.5` in the `sagemaker-huggingface-inference-toolkit` or what do you mean with update?

**Comment by Qingzi-Lan** [comment]
> @philschmid - numpy version needs to be bumped even higher to remove CVEs in inference image. Is it possible to use 1.22.3?

numba 0.55.1(the latest version) which is a dependency of the `sagemaker-huggingface-inference-toolkit` is not compatible with the numpy>=1.22. bases on discussion in numba Github forum, numpy support for numba is about one to two minor version behind the latest numpy version. we will keep numpy<1.22 in this case.

**Comment by Qingzi-Lan** [comment]
container start with the following error, which preventing any command to ran in the container.
```
2022-04-08 22:30:06.345870: W tensorflow/core/profiler/internal/smprofiler_timeline.cc:460] Initializing the SageMaker Profiler.
2022-04-08 22:30:06.346046: W tensorflow/core/profiler/internal/smprofiler_timeline.cc:105] SageMaker Profiler is not enabled. The timeline writer thread will not be started, future recorded events will be dropped.
2022-04-08 22:30:06.374595: W tensorflow/core/profiler/internal/smprofiler_timeline.cc:460] Initializing the SageMaker Profiler.
ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
```

**Comment by philschmid** [comment]
> container start with the following error, which preventing any command to ran in the container.
> 
> ```
> 2022-04-08 22:30:06.345870: W tensorflow/core/profiler/internal/smprofiler_timeline.cc:460] Initializing the SageMaker Profiler.
> 2022-04-08 22:30:06.346046: W tensorflow/core/profiler/internal/smprofiler_timeline.cc:105] SageMaker Profiler is not enabled. The timeline writer thread will not be started, future recorded events will be dropped.
> 2022-04-08 22:30:06.374595: W tensorflow/core/profiler/internal/smprofiler_timeline.cc:460] Initializing the SageMaker Profiler.
> ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> ```

I have build the container locally with 
```
python3 src/main.py --buildspec huggingface/tensorflow/buildspec.yml  --framework huggingface_tensorflow   --image_types inference    --device_types cpu    --py_versions py3
```
and can successfully run it with 
```
docker run -e HF_TASK="automatic-speec-recognition" -e HF_MODEL_ID="hf-internal-testing/tiny-random-wav2vec2" 558105141721.dkr.ecr.us-east-1.amazonaws.com/test:2.6.3-transformers4.17.0-cpu-py38-ubuntu20.04-2022-04-09-08-46-45-pre-push
```

can we see the command which is used to start?

**Comment by philschmid** [comment]
> > > container start with the following error, which preventing any command to ran in the container.
> > > ```
> > > 2022-04-08 22:30:06.345870: W tensorflow/core/profiler/internal/smprofiler_timeline.cc:460] Initializing the SageMaker Profiler.
> > > 2022-04-08 22:30:06.346046: W tensorflow/core/profiler/internal/smprofiler_timeline.cc:105] SageMaker Profiler is not enabled. The timeline writer thread will not be started, future recorded events will be dropped.
> > > 2022-04-08 22:30:06.374595: W tensorflow/core/profiler/internal/smprofiler_timeline.cc:460] Initializing the SageMaker Profiler.
> > > ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> > > ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> > > ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> > > ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> > > ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> > > ```
> > 
> > 
> > I have build the container locally with
> > ```
> > python3 src/main.py --buildspec huggingface/tensorflow/buildspec.yml  --framework huggingface_tensorflow   --image_types inference    --device_types cpu    --py_versions py3
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and can successfully run it with
> > ```
> > docker run -e HF_TASK="automatic-speec-recognition" -e HF_MODEL_ID="hf-internal-testing/tiny-random-wav2vec2" 558105141721.dkr.ecr.us-east-1.amazonaws.com/test:2.6.3-transformers4.17.0-cpu-py38-ubuntu20.04-2022-04-09-08-46-45-pre-push
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > can we see the command which is used to start?
> 
> I used `docker run -it "uri" ./bin/bash`.

Re run it with your command and got this as result 
image

looks okay to me

**Comment by Qingzi-Lan** [comment]
> 

actually, the error happened when start the container with its default start method. `docker run -it "uri"`.
the `./bin/bash` way works. can you try it again and see. thanks

**Comment by Qingzi-Lan** [comment]
> > > > container start with the following error, which preventing any command to ran in the container.
> > > > ```
> > > > 2022-04-08 22:30:06.345870: W tensorflow/core/profiler/internal/smprofiler_timeline.cc:460] Initializing the SageMaker Profiler.
> > > > 2022-04-08 22:30:06.346046: W tensorflow/core/profiler/internal/smprofiler_timeline.cc:105] SageMaker Profiler is not enabled. The timeline writer thread will not be started, future recorded events will be dropped.
> > > > 2022-04-08 22:30:06.374595: W tensorflow/core/profiler/internal/smprofiler_timeline.cc:460] Initializing the SageMaker Profiler.
> > > > ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> > > > ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> > > > ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> > > > ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> > > > ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> > > > ```
> > > 
> > > 
> > > I have build the container locally with
> > > ```
> > > python3 src/main.py --buildspec huggingface/tensorflow/buildspec.yml  --framework huggingface_tensorflow   --image_types inference    --device_types cpu    --py_versions py3
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > and can successfully run it with
> > > ```
> > > docker run -e HF_TASK="automatic-speec-recognition" -e HF_MODEL_ID="hf-internal-testing/tiny-random-wav2vec2" 558105141721.dkr.ecr.us-east-1.amazonaws.com/test:2.6.3-transformers4.17.0-cpu-py38-ubuntu20.04-2022-04-09-08-46-45-pre-push
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > can we see the command which is used to start?
> > 
> > 
> > I used `docker run -it "uri" ./bin/bash`.
> 
> Re run it with your command and got this as result image
> 
> looks okay to me

actually, the error happened when start the container with its default start method. docker run -it "uri".
the ./bin/bash way works. can you try it again and see. thanks

**Comment by philschmid** [comment]
> > > > > container start with the following error, which preventing any command to ran in the container.
> > > > > ```
> > > > > 2022-04-08 22:30:06.345870: W tensorflow/core/profiler/internal/smprofiler_timeline.cc:460] Initializing the SageMaker Profiler.
> > > > > 2022-04-08 22:30:06.346046: W tensorflow/core/profiler/internal/smprofiler_timeline.cc:105] SageMaker Profiler is not enabled. The timeline writer thread will not be started, future recorded events will be dropped.
> > > > > 2022-04-08 22:30:06.374595: W tensorflow/core/profiler/internal/smprofiler_timeline.cc:460] Initializing the SageMaker Profiler.
> > > > > ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> > > > > ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> > > > > ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> > > > > ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> > > > > ERROR - Given model-path /opt/ml/model is not a valid directory. Point to a valid model-path directory.
> > > > > ```
> > > > 
> > > > 
> > > > I have build the container locally with
> > > > ```
> > > > python3 src/main.py --buildspec huggingface/tensorflow/buildspec.yml  --framework huggingface_tensorflow   --image_types inference    --device_types cpu    --py_versions py3
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > and can successfully run it with
> > > > ```
> > > > docker run -e HF_TASK="automatic-speec-recognition" -e HF_MODEL_ID="hf-internal-testing/tiny-random-wav2vec2" 558105141721.dkr.ecr.us-east-1.amazonaws.com/test:2.6.3-transformers4.17.0-cpu-py38-ubuntu20.04-2022-04-09-08-46-45-pre-push
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > can we see the command which is used to start?
> > > 
> > > 
> > > I used `docker run -it "uri" ./bin/bash`.
> > 
> > 
> > Re run it with your command and got this as result image
> > looks okay to me
> 
> actually, the error happened when start the container with its default start method. docker run -it "uri". the ./bin/bash way works. can you try it again and see. thanks

TBH i am not sure how this should work or how the container shouldn't error. if you take a look at the [start_model_server](https://github.com/aws/sagemaker-huggingface-inference-toolkit/blob/2f1fae5cbb3b68299e73cc591c0a912b7cccee29/src/sagemaker_huggingface_inference_toolkit/mms_model_server.py#L52) method which is executed when starting the container there is no way that the container/service starts successful without providing a model or artifacts in `/opt/ml/model` or the "`MODEL_STORE`" 
And there wasn't any change at all to this.

**Review by Qingzi-Lan — APPROVED** [review]

**Review by navinsoni — APPROVED** [review]
